datalad_next/runners/run.py

```diff
@@ -0,0 +1,166 @@
+
+        def arm(self) -> None:
+            self.kill_counter = 0
+            self.armed = True
+
+        def connection_made(self, process: subprocess.Popen) -> None:
+            self.process = process
+            super().connection_made(process)
+
+        def timeout(self, fd: int | None) -> bool:
+            if self.armed:
+                if self.kill_time and self.kill_counter >= self.kill_time:
+                    self.process.kill()
+                    self.kill_time = None
+                if self.terminate_time and self.kill_counter >= self.terminate_time:
+                    self.process.terminate()
+                    self.terminate_time = None
+                self.kill_counter += 1
+
+            # If we set the timeout argument due to a not None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+        cmd: list,
+        protocol_class: type[Protocol],
+        *,
+        cwd: Path | None = None,
+        input: int | IO | bytes | Queue[bytes | None] | None = None,
+        timeout: float | None = None,
+        terminate_time: int | None = None,
+        kill_time: int | None = None,
+) -> Any | Generator:
+
+    introduces_timeout = False
+    if timeout is None:
+        introduces_timeout = True
+        timeout = 1.0
+
+    armed = (
+        False
+        if issubclass(protocol_class, GeneratorMixIn)
+        else True
+    )
+
+    # Create the wrapping class. This is done mainly to ensure that the
+    # termination-related functionality is present in the protocol class that
+    # is used, independent of the actual protocol class that the user passes as
+    # argument.
+    # A side effect of this approach is, that the number of protocol class
+    # definitions is reduced, because the user does not need to define
+    # terminate-capable protocols for every protocol he wants to use.
+    kill_protocol_class = create_kill_wrapper(protocol_class)
+
+    runner = ThreadedRunner(
+        cmd=cmd,
+        protocol_class=kill_protocol_class,
+        stdin=DEVNULL if input is None else input,
+        protocol_kwargs=dict(
+            introduced_timeout=introduces_timeout,
+            terminate_time=terminate_time,
+            kill_time=kill_time,
+            armed=armed,
+            done_future=None,
+            encoding=None
+        ),
+        timeout=timeout,
+        exception_on_error=False,
+        cwd=cwd,
+    )
+    result = runner.run()
+    # We distinguish between a non-generator run, i,e. a blocking run and
+    # a generator run.
+    if not issubclass(protocol_class, GeneratorMixIn):
```

Comment:
Same as `armed is True`?

Reply:
Yes, it is the same, but I think the code is easier to understand this way.